R/reproduction_indices.R

```diff
@@ -116,55 +117,69 @@ lt.reproduction.mortaar_life_table <- function(life_table, fertility_rate = "BA_
 
   indx <- lt.indices(life_table)
 
-  if (fertility_rate == "McO") {
-    fertil_rate <- indx$D0_14_D[[1]] * 7.734 + 2.224
-  } else if (fertility_rate == "BA_linear"){
-    # Linear regression
-    fertil_rate <- indx$p5_19[[1]] * 25.7557 + 2.85273
-  } else if (fertility_rate == "BA_power"){
-    # power fit
-    fertil_rate <- 13.4135 * (indx$p5_19[[1]])**0.374708
-  } else if (fertility_rate == "BA_log"){
-    # logarithmic fit
-    fertil_rate <- log(indx$p5_19[[1]]) * 1.58341 + 9.45601
-  } else if (length(fertility_rate) > 0 & is.numeric(fertility_rate)) {
+  # switch to set fertil_rate
+  if (is.character(fertility_rate)) {
+    switch(fertility_rate,
+      McO = { fertil_rate <- indx$D0_14_D[[1]] * 7.734 + 2.224 },
+      # Linear regression
+      BA_linear = { fertil_rate <- indx$p5_19[[1]] * 25.7557 + 2.85273 },
+      # power fit
+      BA_power = { fertil_rate <- 13.4135 * (indx$p5_19[[1]])**0.374708 },
+      # logarithmic fit
+      BA_log = { fertil_rate <- log(indx$p5_19[[1]]) * 1.58341 + 9.45601 },
+      { stop(paste(
+          "Please choose a valid fertility rate",
+          "(either 'McO', 'BA_linear', 'BA_log' or 'BA_power' or a number)."
+        )) }
+    )
+  } else if (is.numeric(fertility_rate)) {
     fertil_rate <- fertility_rate
   } else {
-    paste0(
-      "Please choose a valid fertility rate (either 'McO', 'BA_linear', 'BA_log' or 'BA_power' or a number)."
-    ) %>%
-      message
+    stop("Please choose a valid fertility rate")
   }
 
   # dependency ratio according to Grupe et al. 2015
   dependency_ratio <- (indx$d0_14 + indx$d60plus) / indx$d20_50
 
   # Survival rates of individuals aged 15--45.
   all_age <- life_table$a %>% cumsum
-  fertil_lx <- life_table$lx[which(all_age >15 & all_age <=45)]
+  fertil_lx <- life_table$lx[all_age >15 & all_age <=45]
   fertil_lx <- c(fertil_lx,100)
 
   # Gross reproductive rate only for females according to Hassan
   R_pot_fem <- fertil_rate * 0.488
 
   # Net reproduction rate after Hassan
-  fertility_perc <- c(10.9, 23.1, 24.1, 19.5, 14.4, 6.2, 1.6)
-  R_0 <- (fertil_lx * fertility_perc * R_pot_fem / 10000) %>% sum
+  #fertility_perc <- c(10.9, 23.1, 24.1, 19.5, 14.4, 6.2, 1.6)
+  #R_0 <- (fertil_lx * fertility_perc * R_pot_fem / 10000) %>% sum
+  R_0 <- NA
```

Comment:
Ok - so here is an implementation error. you can't multiply vectors with different sizes without specifying how exactly this should be done.

To write the testing code I "deactivated" the functionality with `R_0 <- NA`. This has to be fixed. And then the test snaps have to be regenerated.

Reply:
The implementation is unstable because it presupposes 5-year age classes (the default). The Schleswig-dataset uses 7-year-classes for adults. Bad example!
'fertil_lx' and 'fertility_perc' should have the same length and then multiplying (essentially the number of surviving women with the fertility value of this age defined by Hassan) would not be a problem.
Solution 1: Only allow calculations with 5-year-age clases.
Solution 2: Interpolate the Hassan values (f=-148.043 + 19.8072x - 0.786833xx + 0.0126444xxx - 0.73333e-05xxxx gives a good approximation) and then calculate the values for the age-classes used on the fly. Solution 2 would be much more flexible and elegant, of course. I will try something in this direction.
BTW, I found a logic error in my reasoning of putting together 'fertil_lx'. The value '100' denotes that 100% of women living to this age realised their complete fertile phase. But this is already grasped by lx!